toolchain/check/call.cpp

```diff
@@ -342,4 +320,45 @@ auto PerformCall(Context& context, SemIR::LocId loc_id, SemIR::InstId callee_id,
   }
 }
 
+auto PerformCall(Context& context, SemIR::LocId loc_id, SemIR::InstId callee_id,
+                 llvm::ArrayRef<SemIR::InstId> arg_ids) -> SemIR::InstId {
+  // Try treating the callee as a function first.
+  auto callee_function = GetCalleeFunction(context.sem_ir(), callee_id);
+  CARBON_KIND_SWITCH(callee_function.info) {
+    case CARBON_KIND(SemIR::CalleeFunction::Error _): {
+      return SemIR::ErrorInst::InstId;
+    }
+    case CARBON_KIND(SemIR::CalleeFunction::Function fn): {
+      return PerformCallToFunction(context, loc_id, callee_id, fn, arg_ids);
+    }
+    case CARBON_KIND(SemIR::CalleeFunction::Other _): {
+      return PerformCallToOther(context, loc_id, callee_id, arg_ids);
+    }
+
+    case CARBON_KIND(SemIR::CalleeFunction::CppOverloadSet overload): {
+      callee_id = PerformCppOverloadResolution(context, loc_id,
+                                               overload.cpp_overload_set_id,
+                                               overload.self_id, arg_ids);
+      auto overload_result = GetCalleeFunction(context.sem_ir(), callee_id);
+      CARBON_KIND_SWITCH(overload_result.info) {
+        case CARBON_KIND(SemIR::CalleeFunction::Error _): {
+          return SemIR::ErrorInst::InstId;
+        }
+        case CARBON_KIND(SemIR::CalleeFunction::Function fn): {
+          // Preserve the `self` argument from the original callee.
+          CARBON_CHECK(!fn.self_id.has_value());
+          fn.self_id = overload.self_id;
+          return PerformCallToFunction(context, loc_id, callee_id, fn, arg_ids);
+        }
+        case CARBON_KIND(SemIR::CalleeFunction::Other _): {
+          return PerformCallToOther(context, loc_id, callee_id, arg_ids);
```

Comment:
Can we `CARBON_FATAL` here? Overload resolution should either produce a function or an error.

Reply:
(note, I'd incorrectly assumed this was deliberate, that an overload should be allowed to resolve to something else)